# Machine_Learning_MLR_Classifaction_R

## Part 1: Regression 

This part is about regression. Specifically,we will be ``predicting the fuel efficiency`` of a car (in kilometers per litre) based on its characteristics. This is a practical problem as Australia is one of the largest automobile markets in the world; thus, correctly predicting the fuel efficiency is necessary to control emission rates to the environment.

The dataset has many observations and predictors obtained from many retailers for car models available for sale from 2017 to 2020. The target variable is the fuel efficiency of the car measured in kilometers per litre. The higher this value, the better the fuel efficiency of the car. 

In [1]:
# Read the data from students' side
remove(list = ls())
train <- read.csv("RegressionTrain.csv")
test <- read.csv("RegressionTest.csv")

In [ ]:
# PLEASE DO NOT ALTER THIS CODE BLOCK
# Please skip (don't run) this if you are a student
# Read in the data from marking tutors' side (ensure no cheating!)
remove(list = ls())
train <- read.csv("../data/RegressionTrain.csv")
test <- read.csv("../data/RegressionTest.csv")
label <- read.csv("../data/RegressionTestLabel.csv")

### Objective 1 

Fitting a $\textbf{multiple linear model}$ to the fuel efficiency data using the ``train`` dataset. By checking the summary information, which predictors/variables are possibly associated with fuel efficiency (use ``0.05`` significant level), and why? Which ``three predictors/variables`` appear to be the strongest predictors of fuel efficiency, and why? 

In [2]:
# get all column names to see which predictors can be used 
names(train)

# get an overview of the possible values each predictor can take
summary(train)

[1] "Model.Year"              "Eng.Displacement"       
 [3] "No.Cylinders"            "Aspiration"             
 [5] "No.Gears"                "Lockup.Torque.Converter"
 [7] "Drive.Sys"               "Max.Ethanol"            
 [9] "Fuel.Type"               "Comb.FE"

   Model.Year   Eng.Displacement  No.Cylinders    Aspiration    No.Gears     
 Min.   :2017   Min.   :0.900    Min.   : 3.000   N :630     Min.   : 1.000  
 1st Qu.:2017   1st Qu.:2.000    1st Qu.: 4.000   OT:  7     1st Qu.: 6.000  
 Median :2018   Median :3.000    Median : 6.000   SC: 66     Median : 7.000  
 Mean   :2018   Mean   :3.144    Mean   : 5.622   TC:685     Mean   : 6.974  
 3rd Qu.:2019   3rd Qu.:3.625    3rd Qu.: 6.000   TS: 12     3rd Qu.: 8.000  
 Max.   :2020   Max.   :8.000    Max.   :16.000              Max.   :10.000  
 Lockup.Torque.Converter Drive.Sys  Max.Ethanol    Fuel.Type    Comb.FE      
 N: 325                  4:192     Min.   :10.00   G  :642   Min.   : 4.974  
 Y:1075                  A:430     1st Qu.:10.00   GM : 16   1st Qu.: 8.591  
                         F:372     Median :10.00   GP :363   Median : 9.947  
                         P: 37     Mean   :15.11   GPR:379   Mean   :10.447  
                         R:369     3rd Qu.:15.00             3rd

In [3]:
# Comb.FE is our target variable and we use all the other vraibles as predictors
lm.fit <- lm(Comb.FE ~ ., train)

# print overview
summary(lm.fit)


Call:
lm(formula = Comb.FE ~ ., data = train)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.0256 -0.9978 -0.0644  0.7006 11.3941 

Coefficients:
                           Estimate Std. Error t value Pr(>|t|)    
(Intercept)              -1.783e+02  8.587e+01  -2.076  0.03809 *  
Model.Year                9.640e-02  4.255e-02   2.266  0.02363 *  
Eng.Displacement         -1.364e+00  1.025e-01 -13.306  < 2e-16 ***
No.Cylinders              4.644e-02  6.769e-02   0.686  0.49282    
AspirationOT             -3.452e-01  6.352e-01  -0.543  0.58693    
AspirationSC             -9.197e-01  2.282e-01  -4.031 5.85e-05 ***
AspirationTC             -1.303e+00  1.288e-01 -10.111  < 2e-16 ***
AspirationTS             -1.149e+00  4.945e-01  -2.323  0.02035 *  
No.Gears                 -1.307e-01  2.995e-02  -4.364 1.37e-05 ***
Lockup.Torque.ConverterY -8.243e-01  1.117e-01  -7.377 2.78e-13 ***
Drive.SysA               -8.339e-02  1.521e-01  -0.548  0.58356    
Drive.SysF                1.

### Which predictors/variables are possibly associated with fuel efficiency (use 0.05 significant level), and why? 

The predictors:

- Model.Year
- Eng.Displacement
- Aspiration
- No.Gears
- Lockup.Torque.Converter
- Drive.Sys
- Max.Ethanol
- Fuel.Type 

are possibly associated with fuel efficiency. This is since all of the above listed predictors showed a significance level of less than 0.05. A low significance level below 0.05 is indicative of the data rejecting the null hypothesis of no association between the predictor and the target variable. Thus, we could assume that the predictors we outlined above are associated with our target variable, fuel consumption Comb.FE.


### Which three predictors/variables appear to be the strongest predictors of fuel efficiency, and why?

The three predictors:

- AspirationTC with a p-value of < 2e-16
- Eng.Displacement with a p-value of < 2e-16
- Drive.Sys with a p-value of < 2e-16

These three predictors had the lowest p-values, which provides strong evidence to reject the Null-hypothesis. The Null-hypothesis states that there is no association between the predictor variable and the target variable, fuel efficiency. In the case of us failing to reject the Null hypothesis, we can assume that the predictor value is equal to 0. 


### Objective 2

Discussing the effect that the year of manufacture ``(Model.Year)`` variable appears to have on the mean ``fuel efficiency``. Additionally, describe/discuss the effect that the number of gears ``(No.Gears)`` variable has on the mean ``fuel efficiency`` of the car.

### Effect that the year of manufacture (Model.Year) variable appears to have on the mean fuel efficiency.

The predictor Model.Year had a regression coeficent of 9.640e-02, which translates into 0.0964. This means that when the predictor Model.Year increase by one unit, the target variable, fuel efficiency, increases by 0.0964. 

Thus, roughly speaking the the newer a car is the more fuel efficient it is. 

### Effect that the number of gears (No.Gears) variable has on the mean fuel efficiency of the car.

The predictor No.Gears had a regression coeficent of -1.307e-01, which translates into -0.1307.This means that when the predictor No.Gears increase by one unit the target variable, fuel efficiency decreased by -0.1307. 

Thus, roughly speaking the more gears a car has, the less fuel efficeint it is. 

### Objective 3 

Applying the stepwise selection procedure with the $\textbf{BIC}$ penalty to prune out potentially less significant variables. Write down the final regression equation obtained after pruning.

In [4]:
# the idea behind a stepwise selection procedure is to check which of the parameters
# adds value to the model and which don't, thus they will either be added or dropped
# in this case the 'value added' of a parameter in the model is measured using
# the BIC (Bayesian information criterion) penalty 


# number of rows in our train dataframe, represents number of observations 
n = dim(train)[1]

# we can fully automate the stepwise selection procedure using the step() function
# and providing it with:
# the model in use for the current training data set
# trace, in how many steps the information should be printed
# direction, the direction of the testing, which was specified to both
# k, the number of degrees of freedom of the Bayesian information criterion
sw.fit <- step(lm.fit,trace = 0,direction = 'both',k = log(n))

# display the summary of our newly bulit model using the stepwise selection procedure
summary(sw.fit)


Call:
lm(formula = Comb.FE ~ Eng.Displacement + Aspiration + No.Gears + 
    Lockup.Torque.Converter + Drive.Sys + Max.Ethanol, data = train)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.0743 -0.9760 -0.0349  0.6566 11.3971 

Coefficients:
                          Estimate Std. Error t value Pr(>|t|)    
(Intercept)              16.196874   0.282901  57.253  < 2e-16 ***
Eng.Displacement         -1.277173   0.043418 -29.416  < 2e-16 ***
AspirationOT             -0.100081   0.626276  -0.160 0.873060    
AspirationSC             -0.699137   0.213768  -3.271 0.001100 ** 
AspirationTC             -1.144227   0.107302 -10.664  < 2e-16 ***
AspirationTS             -1.122104   0.481471  -2.331 0.019919 *  
No.Gears                 -0.113537   0.029183  -3.891 0.000105 ***
Lockup.Torque.ConverterY -0.825285   0.110202  -7.489 1.23e-13 ***
Drive.SysA                0.035013   0.145617   0.240 0.810020    
Drive.SysF                1.480191   0.166847   8.872  < 2e-16 ***
Drive.SysP 

As the summary above shows only the parameters:

- Eng.Displacement
- AspirationOT
- AspirationSC
- AspirationTC
- AspirationTS
- No.Gears
- Lockup.Torque.ConverterY
- Drive.SysA
- Drive.SysF
- Drive.SysP
- Drive.SysR
- Max.Ethanol

are still in use in the new model sw.fit.

Using the provided Intercept of 16.196874 and coefficients for each parameter we can establish the new final regression equation obtained after pruning.

E[Comb.FE] = 16.20 - 1.28*Eng.Displacement - 0.10*AspirationO - 1.14*AspirationTC - 1.12*AspirationTS - 0.11*No.Gears - 0.83*Lockup.Torque.ConverterY + 0.04*Drive.SysA + 1.48*Drive.SysF - 0.32*Drive.SysP + 0.09*Drive.SysR - 0.01*Max.Ethanol

### Objective 4 

Say we are going to buy a new car and we want to improve the fuel efficiency of our new car, what does this ``BIC model`` suggest we should do?

Based on the above developed BIC model we know that the predictors' Eng.Displacement, AspirationO, AspirationTC, AspirationTS, 1No.Gears, Lockup.Torque.ConverterY, Drive.SysP, Max.Ethanolare are negatively related to our target variable Comb.FE (fuel efficiency). Since fuel efficiency is measured in in kilometers per litre, predictors which are negatively related to fuel efficiency worsen the fuel efficiency of a car.

The predictors' Drive.SysA, Drive.SysF, and Drive.SysR on the other hand are positively related to our target variable Comb.FE (fuel efficiency). Predictors which are positively related to fuel efficiency improve the fuel efficiency of a car.

Generally speaking, to improve the fuel efficiency of a new car we would seek to minimize predictors which negatively related to fuel efficiency and maximize predictors which positively related to fuel efficiency.

More specifically, we would choose a car with the following features:

- a car using Drive.SysF, since it increased fuel efficiency by a factor of 1.4802

- a car with a low number of No.Gears, since it decreased fuel efficiency by a factor of  0.1135

- a car with a low Lockup.Torque.ConverterY, since it decreased fuel efficiency by a factor of 0.8253

- a car with a low AspirationOT, since it decreased fuel efficiency by a factor of 0.1001

- a car with a low Eng.Displacement, since it decreased fuel efficiency by a factor of 1.2772

- a car with a low Max.Ethanol, since it decreased fuel efficiency by a factor of 0.0083

in order to increase fuel efficiency.

### Objective 5

We are looking for a new car to buy to replace an existing car. Using the $\textbf{test}$ dataset to inspect the first car fuel efficiency and see whether it is a good fit for you or not.

In [5]:
# get the first car to test our model since we are tasked to
# use the  test dataset to inspect the first car fuel efficiency 
# and see whether it is a good fit for you or not.
first_car = test[1,]

In [6]:
# now use predict() function to test the predict the 
# mean fuel efficiency
predict(sw.fit,interval = 'confidence',first_car,vcov. = vcov)

fit      lwr      upr     
1 9.287257 9.052956 9.521557

a) Using our BIC model to predict the mean fuel efficiency for this new car. Provide a 95% confidence interval for this prediction. 

Given the above prediction, we can estimate that the expected mean fuel efficiency of a new car should be 9.29 km per litre (9.287257). Moreover, we are 95% confident that this value will fall between a lower interval value of 9.05 km per litre (9.052956) and	a higher interval value of 9.52 km per litre (9.521557).


b) Given that the current car that you own has a mean fuel efficiency of 9.5 km/l (measured over the lifetime of your ownership), does your model (BIC) suggest that the new car will have better fuel efficiency than your current car? Why?

The new car will most likely have better fuel efficiency than your current car. This holds true since we predicted a mean fuel efficiency of 9.29 km per litre (9.287257) and a upper interval value of 9.52 km per litre (9.521557). Thus, using the BIC model, it can be stated that the new car will most likely have a better fuel efficiency than your current car.

### Objective 6 

Save the Planet Regression - Kaggle Competition

Predict the fuel efficiency of motor vehicles

The [link](https://www.kaggle.com/t/0a3c0fc91b074816a6315bb4e9b42602) to the competition.

In [7]:
# Use this function to check the performance of your model
rmse <- function(pred.label, truth.label){
    # Lower is better
    return(sqrt(mean((pred.label - truth.label)^2)))
}

In [8]:
# check which varaibles are continous and thus can be used in a multilinear regression
# get the names of variables
str(train)

'data.frame':	1400 obs. of  10 variables:
 $ Model.Year             : int  2020 2020 2020 2020 2020 2020 2020 2020 2020 2020 ...
 $ Eng.Displacement       : num  1.8 5.2 5.2 3 6.5 5 2 5.2 3.7 5.2 ...
 $ No.Cylinders           : int  4 10 10 6 12 8 4 10 6 12 ...
 $ Aspiration             : Factor w/ 5 levels "N","OT","SC",..: 4 1 1 4 1 3 4 1 1 4 ...
 $ No.Gears               : int  6 7 7 8 7 8 8 7 7 8 ...
 $ Lockup.Torque.Converter: Factor w/ 2 levels "N","Y": 2 1 1 2 1 2 2 2 2 2 ...
 $ Drive.Sys              : Factor w/ 5 levels "4","A","F","P",..: 5 2 2 5 5 2 5 2 5 5 ...
 $ Max.Ethanol            : int  10 15 15 10 10 15 15 15 10 10 ...
 $ Fuel.Type              : Factor w/ 4 levels "G","GM","GP",..: 3 4 4 3 4 3 3 4 4 3 ...
 $ Comb.FE                : num  12.66 7.23 7.23 11.76 5.88 ...


In [9]:
summary(train)

   Model.Year   Eng.Displacement  No.Cylinders    Aspiration    No.Gears     
 Min.   :2017   Min.   :0.900    Min.   : 3.000   N :630     Min.   : 1.000  
 1st Qu.:2017   1st Qu.:2.000    1st Qu.: 4.000   OT:  7     1st Qu.: 6.000  
 Median :2018   Median :3.000    Median : 6.000   SC: 66     Median : 7.000  
 Mean   :2018   Mean   :3.144    Mean   : 5.622   TC:685     Mean   : 6.974  
 3rd Qu.:2019   3rd Qu.:3.625    3rd Qu.: 6.000   TS: 12     3rd Qu.: 8.000  
 Max.   :2020   Max.   :8.000    Max.   :16.000              Max.   :10.000  
 Lockup.Torque.Converter Drive.Sys  Max.Ethanol    Fuel.Type    Comb.FE      
 N: 325                  4:192     Min.   :10.00   G  :642   Min.   : 4.974  
 Y:1075                  A:430     1st Qu.:10.00   GM : 16   1st Qu.: 8.591  
                         F:372     Median :10.00   GP :363   Median : 9.947  
                         P: 37     Mean   :15.11   GPR:379   Mean   :10.447  
                         R:369     3rd Qu.:15.00             3rd

In [10]:
# we reuse the 'brute-force approach' shown in tutorial 10 
# we use log and squaring transformations and include all available continous predictor variables 
fullmod <- lm(Comb.FE ~ . + .*. + log(Model.Year) + log(Eng.Displacement) + log(No.Cylinders) + log(No.Gears) + log(Max.Ethanol) 
              + I(Model.Year^2) + I(Eng.Displacement^2) + I(No.Cylinders^2) + I(No.Gears^2) + I(Max.Ethanol^2)  , 
              data=train)


# lastly we use the step function which "fine tunes" the model and only keeps the parmeters that 
# dd value to the model
# this avoids unnecessary complexity 
fin.mod <- step(fullmod, k = log(1400), trace=0, direction="both")


In [11]:
# model rmse comparison - generate predicted values with each model using predict function with the train dataset
pred.label.sw.trans <- predict(fin.mod, train[1:9])
pred.label.sw <- predict(sw.fit, train[1:9])
pred.label.lm<- predict(lm.fit, train[1:9])

Warning message in predict.lm(fin.mod, train[1:9]):
“prediction from a rank-deficient fit may be misleading”

In [12]:
# rmse returned by the rmse function with train data set
paste('simple multiple regression rmse: ',rmse(pred.label = pred.label.lm, truth.label = train$Comb.FE))
paste('stepwise multiple regression rmse: ',rmse(pred.label = pred.label.sw, truth.label = train$Comb.FE))
paste('complex stepwise multiple regression rmse: ',rmse(pred.label = pred.label.sw.trans, truth.label = train$Comb.FE))

[1] "simple multiple regression rmse:  1.58763251824734"

[1] "stepwise multiple regression rmse:  1.5973496190278"

[1] "complex stepwise multiple regression rmse:  1.3037409137109"

In [13]:
# generate predicted Comb.FE using the final model 
pred.label <- predict(fin.mod, test)

Warning message in predict.lm(fin.mod, test):
“prediction from a rank-deficient fit may be misleading”

In [14]:
# PLEASE DO NOT ALTER THIS CODE BLOCK
# put this label in a csv file to commit to the Leaderboard
write.csv(data.frame("RowIndex" = seq(1, length(pred.label)), "Prediction" = pred.label),  
          "RegressionPredictLabel.csv", row.names = F)


In [ ]:
## PLEASE DO NOT ALTER THIS CODE BLOCK
## Please skip (don't run) this if you are a student
## For teaching team use only
RMSE.fin <- rmse(pred.label, label$Label)
cat(paste("RMSE is", RMSE.fin))

The thought process that went into building this model:

We created a logistic regression model which, based on inputs such as education level, gender, etc., seeks to determine whether someone is wealthy or not. The process of building a logistic regression model is similar to  multilinear regression model is a continuous trial and error process of making gradual adjustments to the model which enhances its predictive ability.

Every multilinear regression model has a base format of:

$ \widehat{y} =  \beta_0 + \sum_{j=i}^{p}\beta_j*x_{i,j} $

Given this format, we seek to find the best predictors and transformations that can be applied to the predictors to enhance our model's performance. Performance in this case is measured using mean squared error.

Our first step was to identify all variables that can be used as predictor variables for a multilinear regression model. This was done using the str(train) command. Moreover, using the summary(train) command we assess the statistical characteristics of potential predictor variables. After that, we insert all identified variables that can be used as predictor variables into the model. Additionally, we reuse the 'brute-force approach' shown in tutorial 10 - adding predictors after applying logarithms and polynomial degree of 2, as well as interaction between all the predictor variables. In this 'brute-force approach' we 'throw all possible model enhancements at the wall and see what sticks'.

After creating such a complex model with many predictors we need to check which of these predictors add value to the model (increase its accuracy). Thus we apply the stepwise selection procedure. The idea behind a stepwise selection procedure is to check which of the parameters adds value to the model and which don't, thus they will either be added or dropped. In this case, the 'value-added' of a parameter in the model is measured using the BIC (Bayesian information criterion) penalty. In R we can fully automate the stepwise selection procedure using the step() function.

After creating this model, I compared the RMSE of the simple multiple regression built at Q1 (lm.fit), the multiple regression with stepwise regression at Q3(sw.fit), and my new model with brute-force approach of stepwise regression. The returned RMSE of the new model built had a significantly lower RMSE compared to two other models from above. This was also shown by the rmse score from Kaggle with the test data set. Therefore, I decided to use the new model as my final model. 

# Part 2: Classification

In this part, we are going to work with "Census Income Dataset" which was originally donated by Ronny Kohavi and Barry Becker to UCI (University of California, Irvine) in 1996. This is a trimmed dataset used for machine learning students to study classification. 

This dataset has collected over 40,000 records (we excluded some data in our version) regarding personal yearly income with 12 attributes (predictors). The attributes comprise many aspects of a person that may contribute to the yearly income. You can use summary() function to obtain the attributes information. Your prediction task is to determine whether a person makes over 50K a year.

We have splitted the dataset into a trainning and a testing set. There are 27245 records in the training set while 13631 records in the testing set. Besides the 12 predictors, there is one more column named Salary indicating whether a person's yearly income is over 50K. The label information is a seperated file for the testing set and will be used by us to asess your performance later. Note the label TRUE means an individual's yearly salary exceeds 50K while FALSE means an individual's yearly salary is under 50K.

In [15]:
# Read the data from students' side
remove(list = ls())
train <- read.csv("ClassTrain.csv")
test  <- read.csv("ClassTest.csv")

In [ ]:
## PLEASE DO NOT ALTER THIS CODE BLOCK
# Please skip (don't run) this if you are a student
# Read in the data from marking tutors' side (ensure no cheating!)
remove(list = ls())
train <- read.csv("../data/ClassTrain.csv")
test  <- read.csv("../data/ClassTest.csv")
label <- read.csv("../data/ClassTestLabel.csv")

### Objective  1

Fit a $\textbf{Generalized Linear Model (Logistic Regression)}$ to predict level of income (salary) $\left(\;\geq 50\;\text{K, or } <50\;\text{K}\;\right)$ using the ``train`` dataset. Using the results of fitting this model, which predictors do you think are possibly associated with the level of Salary (use ``0.05`` significant level), and why? Which ``three variables`` appear to be the strongest predictors of salary, and why? 

Furthermore, you can see that you have much more predictors in this part than in the ``linear model`` from Part 1 $\Rightarrow$ manually checking information is counterproductive. Thus, please write a function to automate these processes $\textbf{(1)}$ selecting important feature against 0.05 threshold and $\textbf{(2)}$ Selecting three most important features.

In [16]:
# get an overview of the variables avaiable as parameters and their values
summary(train)

      Age                   WorkClass      FinalWeight             Education   
 Min.   :17.00   Federal-gov     :  874   Min.   :  13769   HS-grad     :9152  
 1st Qu.:28.00   Local-gov       : 1947   1st Qu.: 115896   Some-college:6240  
 Median :37.00   Private         :19936   Median : 176814   Bachelors   :4609  
 Mean   :38.43   Self-emp-inc    :  990   Mean   : 187061   Masters     :1483  
 3rd Qu.:47.00   Self-emp-not-inc: 2294   3rd Qu.: 234096   Assoc-voc   :1228  
 Max.   :90.00   State-gov       : 1204   Max.   :1484705   11th        : 946  
                                                            (Other)     :3587  
               MaritalStatus             Occupation           Relationship  
 Divorced             : 3962   Exec-managerial:3731   Husband       :11346  
 Married-civ-spouse   :12716   Prof-specialty :3687   Not-in-family : 7052  
 Married-spouse-absent:  229   Craft-repair   :3662   Other-relative:  628  
 Never-married        : 8796   Adm-clerical   :3439 

In [17]:
# build a Generalized Linear Model using the glm function 
# Generalized Linear Models are in use to determine associations and realtionships and 
# can be used for a variaty of models such Linear Regressions and Logistic Regression
glm.fit <- glm(Salary ~., family=binomial, data=train)

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”

In [18]:
# then print  summary of our Generalized Linear Model to check which predictors are statistically significant
summary(glm.fit)


Call:
glm(formula = Salary ~ ., family = binomial, data = train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-5.1013  -0.5296  -0.1926   0.0276   3.4349  

Coefficients:
                                     Estimate Std. Error z value Pr(>|z|)    
(Intercept)                        -7.614e+00  4.525e-01 -16.826  < 2e-16 ***
Age                                 2.626e-02  1.779e-03  14.762  < 2e-16 ***
WorkClassLocal-gov                 -7.214e-01  1.168e-01  -6.179 6.46e-10 ***
WorkClassPrivate                   -4.734e-01  9.693e-02  -4.884 1.04e-06 ***
WorkClassSelf-emp-inc              -2.974e-01  1.283e-01  -2.317 0.020506 *  
WorkClassSelf-emp-not-inc          -9.994e-01  1.139e-01  -8.772  < 2e-16 ***
WorkClassState-gov                 -7.757e-01  1.294e-01  -5.996 2.03e-09 ***
FinalWeight                         7.896e-07  1.822e-07   4.334 1.46e-05 ***
Education11th                       6.909e-02  2.201e-01   0.314 0.753589    
Education12th            

In [19]:
# Task: please write a function to automate these processes selecting important feature against 0.05

# function that only return statistically significant predictors called return_signif_predictors
# fucntion has two inputs model, where we input the model in use and threshold, that is used to decide
# which p-value we used to determine the signifigance of a factor
return_signif_predictors <- function(model,threshold)
    
{
    
    # we seek to extract the predictors that have p-values below our threshold 
    # code to extract predictors based on p-values taken from
    # https://stackoverflow.com/questions/23838937/extract-pvalue-from-glm
    # user: R. Prost   accessed: 06/11/2020
    get_p_value <- coef(summary(model))[,'Pr(>|z|)']
    
    # keep all the predictors that have p-values below our threshold 
    p_value_below_threshold <- get_p_value[get_p_value < threshold]
    
    # don't keep the (Intercept) value since that is not a predictor we can use
    p_value_below_threshold <- p_value_below_threshold[labels(p_value_below_threshold)!='(Intercept)']
    
    # selecting three most important predictors based on their coefficient 
    # useing rank() function to rank/order predictors and then select the top 3
    top_three <- labels(p_value_below_threshold[rank(p_value_below_threshold) %in%1:3])
    
    # finally, we print all the results of our function
    print(paste('The following variables are possibly associated with our target variable the level of Salary:',
               paste(labels(p_value_below_threshold),collapse = ', ')))
    
    print('This is since they all had a p value lower than 0.05, which provides strong evidence to reject the Null-hypothesis.')
    
    print(paste('The strongest three predictors of salary were the variables:',paste(top_three,collapse=', ')))
    
}

In [20]:
# call the function return_signif_predictors with two inputs 
return_signif_predictors(model = glm.fit, threshold = 0.05)

[1] "The following variables are possibly associated with our target variable the level of Salary: Age, WorkClassLocal-gov, WorkClassPrivate, WorkClassSelf-emp-inc, WorkClassSelf-emp-not-inc, WorkClassState-gov, FinalWeight, Education7th-8th, EducationAssoc-acdm, EducationAssoc-voc, EducationBachelors, EducationDoctorate, EducationHS-grad, EducationMasters, EducationProf-school, EducationSome-college, MaritalStatusMarried-civ-spouse, MaritalStatusNever-married, OccupationExec-managerial, OccupationFarming-fishing, OccupationHandlers-cleaners, OccupationMachine-op-inspct, OccupationOther-service, OccupationProf-specialty, OccupationProtective-serv, OccupationSales, OccupationTech-support, RelationshipNot-in-family, RelationshipOwn-child, RelationshipWife, RaceAsian-Pac-Islander, RaceWhite, GenderMale, CapitalGain, CapitalLoss, HoursWork"
[1] "This is since they all had a p value lower than 0.05, which provides strong evidence to reject the Null-hypothesis."
[1] "The strongest three pred

### Objective 2 

Firstly, use the model created in the previous question to predict for the labels of the $\textbf{train}$ data. Consequently, our objective is to compare this ``predict.label`` with the ``truth.label`` from the $\textbf{test}$ data. However, as we don't know the $\textbf{test}$ label, we have to estimate model performance using $\textbf{train}$ data at this moment.

Secondly, since our objective is to estimate the performance of this model in making correct predictions; thus, this question also asks you to explore different [performance metrics](https://en.wikipedia.org/wiki/Precision_and_recall) for classification models. The metrics we will use are $\textbf{Accuracy, Precision, Recall, and F1 Score}$, please create a function to calculate these value and print them out properly using the given structure.

In [21]:
# Apply your previous model to perform prediction, keep type = "response"
# Don't worry if you receive some warnings, they are benign
# using the factor function we turn a probability value into a Flase or True Boolean variable
# this is our predictor label, whether someone is wealthy or not
predict.label <- factor(predict(glm.fit, type="response") > 0.5, c(F,T))

# truth label from train data
truth.label <- train$Salary

In [22]:
# Model statistics function
mod.stat <- function(predict.label, truth.label){
    # instantiate the variables
    accuracy <- NULL
    precision <- NULL
    recall <- NULL
    F1 <- NULL
    
    ##############################
    #Your calculatation here
    
    # reuseing code from tutorial week 10 r file
    
    # create a table that summarizes the true & false negatives as well as the true & false positives
    T = table(truth.label, predict.label)
    
    # calculate the accuracy of our model 
    accuracy <- round(mean(truth.label==predict.label), digits = 4)
    
    # calculate the recall of our model 
    recall <- round(T[2,2]/(T[2,1] + T[2,2]), digits = 4)
    
    # calculate the precision of our model
    precision <- round(T[2,2]/(T[2,2]+T[1,2]), digits = 4)
    
    # calculate the F1 of our model
    F1 <- round(2/(recall^(-1)+precision^(-1)), digits = 4)
    
    ##############################
    
    # print the results of our calaculations
    return(list("accuracy" = accuracy, "precision" = precision, "recall" = recall, "fscore" = F1))
}

In [23]:
# Run the function to get statistics, provide description/discussion after this
mod.stat(predict.label, truth.label)

$accuracy
[1] 0.8452

$precision
[1] 0.7395

$recall
[1] 0.6107

$fscore
[1] 0.669

Result Discussion:

The created classification model seeks to predict whether someone is wealthy or not.
The calculated performance metrics for our classification model show that our model has a moderate level of precision, 0.7395 or 74%, and a relatively low level of recall, 0.6107 or 61%. It makes sense for this model to enhance the precision of the model, the accuracy of the model given that our target is to determine if someone is wealthy. Given the high precision of the model, we expected to be able to more accurately predict that someone makes more than 50k per year (is wealthy). The trade off with focusing on the precision of the model is that the recall is likely to be lower. The model's overall F-score is in between these values at 0.669 or 67%.

$accuracy
0.8452
$precision
0.7395
$recall
0.6107
$fscore
0.669

### Objective 3 

Using the stepwise selection procedure with the $\textbf{BIC}$ penalty to prune out potentially unimportant variables. Checking the performance of your model using the created ``mod.stat()`` function, please give your discussion as how this model is compared with the ``glm.fit``(you can run the ``mod.stat()`` function for this as well if you want to).

In [24]:
# Setting to suppress warnings
options(warn=-1)

# Fit a stepwise model
# using the step() function we optimize our model and prune out potentially unimportant variables
sw.fit <- step(glm.fit, k = log(27245), direction = 'both', trace =0)

# Setting to suppress warnings
options(warn=0)

# print the summary  of our optimized model to enhance our understand of it
summary(sw.fit)


Call:
glm(formula = Salary ~ Age + WorkClass + FinalWeight + Education + 
    MaritalStatus + Occupation + Relationship + Gender + CapitalGain + 
    CapitalLoss + HoursWork, family = binomial, data = train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-5.0961  -0.5291  -0.1936   0.0279   3.4393  

Coefficients:
                                     Estimate Std. Error z value Pr(>|z|)    
(Intercept)                        -7.130e+00  3.929e-01 -18.146  < 2e-16 ***
Age                                 2.644e-02  1.778e-03  14.869  < 2e-16 ***
WorkClassLocal-gov                 -7.159e-01  1.163e-01  -6.154 7.54e-10 ***
WorkClassPrivate                   -4.588e-01  9.626e-02  -4.766 1.88e-06 ***
WorkClassSelf-emp-inc              -2.776e-01  1.278e-01  -2.173  0.02977 *  
WorkClassSelf-emp-not-inc          -9.857e-01  1.133e-01  -8.703  < 2e-16 ***
WorkClassState-gov                 -7.653e-01  1.290e-01  -5.930 3.02e-09 ***
FinalWeight                         7.

In [25]:
# Making prediction using train data and view the statistics
# using the factor function we turn a probability value into a Flase or True Boolean variable
# this is our predictor label, whether someone is wealthy or not
predict.label.sw <- factor(predict(sw.fit, type="response")>0.5, c(F,T))

# Only run the below if you have labels, in your submission, this must be UNCOMMENTED
# checking the performance of your model using the created mod.stat() function
mod.stat(predict.label.sw, truth.label)

$accuracy
[1] 0.8451

$precision
[1] 0.7396

$recall
[1] 0.6104

$fscore
[1] 0.6688

The newly created classification model, sw.fit, seeks to predict whether someone is wealthy or not.
The new 'optimized model', sw.fit, has not improved performance compared to the old model glm.fit.
The new model, sw.fit, has almost identitcal accuracy, precision, recall, and fscore values. Thus regardless of the pruning out of potentially unimportant variables, the model performance did not increase (see below).

OLD model: glm.fit
$accuracy
0.8452
$precision
0.7395
$recall
0.6107
$fscore
0.669

NEW model: sw.fit
$accuracy
0.8451
$precision
0.7396
$recall
0.6104
$fscore
0.6688

### Objective 4

Rich or Not Classification Task - Kaggle Competition

Predict whether a person makes more or less than $50k

The [link](https://www.kaggle.com/t/1bdebc96607742dbaf47ab36cd3ae421) to the competition

In [27]:
# Assess the statistical characteristics of the predictor variables
summary(train)

      Age                   WorkClass      FinalWeight             Education   
 Min.   :17.00   Federal-gov     :  874   Min.   :  13769   HS-grad     :9152  
 1st Qu.:28.00   Local-gov       : 1947   1st Qu.: 115896   Some-college:6240  
 Median :37.00   Private         :19936   Median : 176814   Bachelors   :4609  
 Mean   :38.43   Self-emp-inc    :  990   Mean   : 187061   Masters     :1483  
 3rd Qu.:47.00   Self-emp-not-inc: 2294   3rd Qu.: 234096   Assoc-voc   :1228  
 Max.   :90.00   State-gov       : 1204   Max.   :1484705   11th        : 946  
                                                            (Other)     :3587  
               MaritalStatus             Occupation           Relationship  
 Divorced             : 3962   Exec-managerial:3731   Husband       :11346  
 Married-civ-spouse   :12716   Prof-specialty :3687   Not-in-family : 7052  
 Married-spouse-absent:  229   Craft-repair   :3662   Other-relative:  628  
 Never-married        : 8796   Adm-clerical   :3439 

In [28]:
# An example would be use the previous model as your final one
# we can re-use our earlier created model sw.fit
fin.mod <- sw.fit

In [31]:
# Getting the predict label for the TEST data
# using the factor function we turn a probability value into a Flase or True Boolean variable
# this is our predictor label, whether someone is wealthy or not
pred.label <- factor(predict(sw.fit,train[1:12], type="response")>0.5, c(F,T))

In [32]:
mod.stat(pred.label, train$Salary)

$accuracy
[1] 0.8451

$precision
[1] 0.7396

$recall
[1] 0.6104

$fscore
[1] 0.6688

In [43]:
# check model performance using threshold of 0.4
pred.label <- factor(predict(sw.fit,train[1:12], type="response")>0.4, c(F,T))
mod.stat(pred.label, train$Salary)

$accuracy
[1] 0.8415

$precision
[1] 0.6853

$recall
[1] 0.7054

$fscore
[1] 0.6952

In [42]:
# PLEASE DO NOT ALTER THIS CODE BLOCK
# Use this csv file to commit to the leaderboard
write.csv(data.frame("RowIndex" = seq(1, length(pred.label)), "Prediction" = pred.label),  
          "ClassPredictLabel.csv", row.names = F)

In [ ]:
## PLEASE DO NOT ALTER THIS CODE BLOCK
## Please skip (don't run) this if you are a student
## For teaching team use only
source("../data/modassess.r")
model.perf <- mod.stat.test(pred.label,label$Label)
print(model.perf)

Result Discussion:

The thought process that went into building this model:

We created a logistic regression model which, based on inputs such as education level, gender, etc., seeks to determine whether someone is wealthy or not. The first steps in the process of building a logistic regression model are similar to the steps taken in building a multilinear regression model. Using the glm() function we create a Generalised Linear Model which can be used to determine associations and relationships for a variety of models such Linear and Logistic Regressions. 

This Generalised Linear Model has a base format of:

$ v =  \beta_0 + \sum_{j=i}^{p}\beta_j*x_{i,j} $

In this case we reused the earlier created model sw.fit. To build that model we sought to find the best predictors and transformations that can be applied to the predictors to enhance our model's performance. Our first step was to identify all variables that can be used as a predictor variable and incorporate them into the model. After creating a Generalised Linear Model using all available predictors we need to check which of these predictors add value to the model. We apply the stepwise selection procedure where we check which of the parameters add value to the model and which don't. Thus parameters will either be added or dropped. In this case, the 'value-added' of a parameter in the model is measured using the BIC (Bayesian information criterion) penalty. In R we can fully automate the stepwise selection procedure using the step() function.


At this point the steps to turn a 'regular' regression model into a logistic regression begin. The returned values from glm function are linear predictors, which should be 'normalised' to a value between 0 to 1. To do so, the predict() variable in combination of type attribute specified as 'response' returns the linear predictors after applying sigmoid function $ S(x) = \frac{1}{1 + e^{-v}}$. 

After normalising the linear predictors, we set a threshold to apply the binary classification. 

Using the a combination of the factor() and predict() function we use the earlier constructed sw.fit model to make True False predictions on whether someone is wealthy or not. This binary classification is based on the Sigmoid function which takes the form of:

$ S(x) = \frac{1}{1 + e^{-v}}$

To look for an optimal threshold for binary classification, I tried different threshold level between 0.35 to 0.6. However, the accuracy and f-score did not drastically improve based on the difference in threshold level, and I needed to compromise between the f-score and accuracy when changing the threshold. Therefore, the most optimal threshold for the model was defined as 0.5. 